pdf_to_json/src/services/tei_to_json.py

```diff
@@ -0,0 +1,85 @@
+"""
+This module provides utilities for converting XML documents in the TEI format to
+JSON format. It includes functionality to parse XML elements recursively and
+to save the resulting JSON representation to a specified file location.
+"""
+
+from lxml import etree
+import json
+from typing import Union, Dict, Any
+
+
+def xml_to_dict(element: etree._Element) -> Union[Dict[str, Any], str]:
+  """
+  Recursively converts an XML element and its children into a dictionary.
+
+  Args:
+      element: The XML element to convert.
+
+  Returns:
+      A dictionary representation of the input XML element or a string if the
+      element contains only text.
+  """
+  if not element.getchildren():
+    if element.text:
+      return element.text.strip()
+    else:
+      return {k: v for k, v in element.items()}
+  else:
+    child_dict: Dict[str, Union[Dict, list]] = {}
+    for child in element:
+      child_result = xml_to_dict(child)
+      if child.tag in child_dict:
+        if not isinstance(child_dict[child.tag], list):
+          child_dict[child.tag] = [child_dict[child.tag]]
+        child_dict[child.tag].append(child_result)
+      else:
+        child_dict[child.tag] = child_result
+    if element.items():
+      child_dict["@attributes"] = {k: v for k, v in element.items()}
+    return child_dict
+
+
+def tei_to_json(input_tei_path: str, output_json_path: str) -> None:
```

Comment:
https://github.com/allenai/s2orc-doc2json

I think this library does exactly what you are trying to do here. Can you reuse their work?